test/CMakeLists.txt

```diff
@@ -31,20 +31,28 @@ include_directories(${CMAKE_CURRENT_SOURCE_DIR}/..)
 add_executable(utf8_test
   main.c
   test.c
-  test.cpp
 )
 
+add_executable(utf8_test_cpp14
+    test.cpp
+)
+target_compile_features(utf8_test_cpp14 PRIVATE cxx_std_14)
+add_executable(utf8_test_cpp20
+    test.cpp
+    )
+target_compile_features(utf8_test_cpp20 PRIVATE cxx_std_20)
+
 if("${CMAKE_C_COMPILER_ID}" STREQUAL "GNU")
   set_source_files_properties(test.c test.cpp PROPERTIES
     COMPILE_FLAGS "-Wall -Wextra -Werror"
   )
 elseif("${CMAKE_C_COMPILER_ID}" STREQUAL "Clang")
   set_source_files_properties(test.c test.cpp PROPERTIES
-    COMPILE_FLAGS "-Wall -Wextra -Weverything -Werror"
+    COMPILE_FLAGS "-Wall -Wextra -Weverything -Werror -Wno-c++98-compat"
   )
 elseif("${CMAKE_C_COMPILER_ID}" STREQUAL "MSVC")
   set_source_files_properties(test.c test.cpp PROPERTIES
-    COMPILE_FLAGS "/Wall /WX /wd4514"
+    COMPILE_FLAGS "/Wall /WX /wd4514 /Zc:__cplusplus"
```

Comment:
Why are you defining __cplusplus even for C source file?

Reply:
This flag tells MSVC to actually use the real value of `__cplusplus`. Without it, it will default to C++98
It doesn't matter for the C file, as it will stay undefined
See https://devblogs.microsoft.com/cppblog/msvc-now-correctly-reports-__cplusplus/